components/TextFields/src/MDCTextInputController.h

```diff
@@ -0,0 +1,200 @@
+
+/**
+ Controls when the character count will be shown and therefore whether character counting determines
+ error state.
+
+ Default is UITextFieldViewModeNever.
+ */
+@property(nonatomic, assign) UITextFieldViewMode characterCountViewMode NS_SWIFT_NAME(characterMode)
+    ;
+
+/**
+ The color used to denote error state in the underline, the errorText's label, the placeholder and
+ the character count label.
+
+ Default is red.
+ */
+@property(nonatomic, nullable, strong) UIColor *errorColor UI_APPEARANCE_SELECTOR;
+
+/**
+ The color applied to the placeholder when floating. However, when in error state, it will be
+ colored with the error color.
+
+ Default is black with Material Design hint text opacity.
+ */
+@property(nonatomic, nullable, strong)
+    UIColor *floatingPlaceholderColor NS_SWIFT_NAME(floatingColor) UI_APPEARANCE_SELECTOR;
+
+/**
+ The scale of the the floating placeholder label in comparison to the inline placeholder specified
+ as a value from 0.0 to 1.0.
+
+ Default is 0.75.
+ */
+@property(nonatomic, nullable, strong)
+    NSNumber *floatingPlaceholderScale NS_SWIFT_NAME(floatingScale) UI_APPEARANCE_SELECTOR;
+
+/**
+ Text displayed in the leading underline label.
+
+ This text should give context or instruction to the user. If error text is set, it is
+ not shown.
+ */
+@property(nonatomic, nullable, strong) IBInspectable NSString *helperText NS_SWIFT_NAME(helper);
+
+/**
+ The color applied to the placeholder when inline (not floating).
+
+ Default is black with Material Design hint text opacity.
+ */
+@property(nonatomic, nullable, strong) UIColor *inlinePlaceholderColor NS_SWIFT_NAME(inlineColor)
+    UI_APPEARANCE_SELECTOR;
+
+/*
+ Indicates whether the alert contents should automatically update their font when the device’s
+ UIContentSizeCategory changes.
+
+ This property is modeled after the adjustsFontForContentSizeCategory property in the
+ UIConnectSizeCategoryAdjusting protocol added by Apple in iOS 10.0.
+
+ Default value is NO.
+ */
+@property(nonatomic, readwrite, setter=mdc_setAdjustsFontForContentSizeCategory:)
+    BOOL mdc_adjustsFontForContentSizeCategory UI_APPEARANCE_SELECTOR;
+
+/** The behavioral style applied to the text input. */
+@property(nonatomic, assign)
+    MDCTextInputPresentationStyle presentationStyle NS_SWIFT_NAME(presentation);
+
+/** The text input the controller is affecting. */
+@property(nonatomic, nullable, strong) UIView<MDCTextInput> *textInput NS_SWIFT_NAME(input);
+
+/**
+ Controls when the underline will be shown.
+
+ The underline is an overlay that can be hidden depending on the editing state of the input text.
+
+ Default is UITextFieldViewModeAlways.
+ */
+@property(nonatomic, assign) UITextFieldViewMode underlineViewMode NS_SWIFT_NAME(underlineMode)
+    UI_APPEARANCE_SELECTOR;
+
+/**
+ Convenience init. Never fails.
+
+ @param input An MDCTextInput this controller will manage.
+ */
+- (nonnull instancetype)initWithTextInput:(nullable UIView<MDCTextInput> *)input
```

Comment:
Yes. See above comment. If the text input is deallocated because the view hierarchy stops existing, that shouldn't cause the controller to make crashes.